src/main/kotlin/it/pagopa/wallet/cdc/PaymentWalletsLogEventsStream.kt

```diff
@@ -44,22 +51,49 @@ class PaymentWalletsLogEventsStream(
                                 Aggregation.project(changeStreamOptionsConfig.project)
                             )
                         )
-                        .resumeAt(Instant.now())
+                        .resumeAt(redisResumePolicyService.getResumeTimestamp())
                         .build(),
                     BsonDocument::class.java
                 )
-                .flatMap {
-                    Mono.defer {
-                            walletPaymentCDCEventDispatcherService.dispatchEvent(
-                                it.raw?.fullDocument?.toBsonDocument()
-                            )
-                        }
-                        .onErrorResume {
-                            logger.error("Error during event handling : ", it)
-                            Mono.empty<BsonDocument>()
-                        }
-                }
+                // Process the elements of the Flux
+                .flatMap { processEvent(it) }
+                // Save resume token every n emitted elements
+                .index()
+                .flatMap { saveCdcResumeToken(it) }
+                .doOnError { logger.error("Error listening to change stream: ", it) }
 
         return flux
     }
+
+    private fun processEvent(event: ChangeStreamEvent<BsonDocument>): Mono<BsonDocument> {
+        return Mono.defer {
+                walletPaymentCDCEventDispatcherService.dispatchEvent(
+                    event.raw?.fullDocument?.toBsonDocument()
+                )
+            }
+            .onErrorResume {
+                logger.error("Error during event handling : ", it)
+                Mono.empty()
+            }
+    }
+
+    private fun saveCdcResumeToken(
+        changeEventFluxIndex: Tuple2<Long, BsonDocument>
```

Comment:
given the above refactoring we can use explicit parameter here instead of tuple one